File: rest-common/src/main/java/org/chronopolis/rest/api/ServiceGenerator.java
package org.chronopolis.rest.api;

/**
 * Interface to create the various services for use when communicating with the
 * Ingest server
 *
 * todo: Service -> API
 *
 * @author shake
 */
public interface ServiceGenerator {

    /**
     * Create a {@link BagService} for querying Bags in the Ingest Server
     *
     * @return a new BagService
     */
    BagService bags();

    /**
     * Create a {@link TokenService} for querying Bag Tokens in the Ingest Server
     *
     * @return a new TokenService
     */
    TokenService tokens();

    /**
     * Create a {@link RepairService} for querying Repairs in the Ingest Server
     *
     * @return a new RepairService
     */
    RepairService repairs();

    /**
     * Create a {@link StagingService} for querying Bag Staging information in the Ingest Server
     *
     * @return a new StagingService
     */
    StagingService staging();

    /**
     * Create a {@link DepositorAPI} for querying Depositors in the Ingest Server
     *
     * @return a new DepositorAPI
     */
    DepositorAPI depositorAPI();

    /**
     * Create a {@link StorageService} for querying Storage Regions in the Ingest Server
     *
     * @return a new StorageService
     */
    StorageService storage();

    /**
     * Create a {@link ReplicationService} for querying Replications the Ingest Server
     *
     * @return a new ReplicationService
     */
    ReplicationService replications();

}

File: rest-models/src/main/kotlin/org/chronopolis/rest/api/ServiceGenerator.kt
package org.chronopolis.rest.api

/**
 * Interface to create the various services for use when communicating with the
 * Ingest server
 *
 * @author shake
 */
interface ServiceGenerator {

    /**
     * Create a [BagService] for querying Bags in the Ingest Server
     *
     * @return a new BagService
     */
    fun bags(): BagService

    /**
     * Create a [FileService] for querying Bags in the Ingest Server
     *
     * @return a new BagService
     */
    fun files(): FileService

    /**
     * Create a [TokenService] for querying Bag Tokens in the Ingest Server
     *
     * @return a new TokenService
     */
    fun tokens(): TokenService

    /**
     * Create a [RepairService] for querying Repairs in the Ingest Server
     *
     * @return a new RepairService
     */
    fun repairs(): RepairService

    /**
     * Create a [StagingService] for querying Bag Staging information in the Ingest Server
     *
     * @return a new StagingService
     */
    fun staging(): StagingService

    /**
     * Create a [DepositorService] for querying Depositors in the Ingest Server
     *
     * @return a new DepositorService
     */
    fun depositors(): DepositorService

    /**
     * Create a [StorageService] for querying Storage Regions in the Ingest Server
     *
     * @return a new StorageService
     */
    fun storage(): StorageService

    /**
     * Create a [ReplicationService] for querying Replications the Ingest Server
     *
     * @return a new ReplicationService
     */
    fun replications(): ReplicationService

}
